File: packages/pluto/src/clients/aws/kvstore.dynamodb.ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { PutCommand, GetCommand, DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { IKVStoreClient, KVStore, KVStoreOptions } from "../../kvstore";
import { genResourceId } from "@plutolang/base/utils";
import { genAwsResourceName } from "./utils";

/**
 * Implementation of KVStore using AWS DynamoDB.
 */
export class DynamoKVStore implements IKVStoreClient {
  private readonly id: string;
  private readonly tableName: string;

  private client: DynamoDBClient;
  private docClient: DynamoDBDocumentClient;

  constructor(name: string, opts?: KVStoreOptions) {
    this.id = genResourceId(KVStore.fqn, name);
    this.tableName = genAwsResourceName(this.id);

    this.client = new DynamoDBClient();
    this.docClient = DynamoDBDocumentClient.from(this.client);
    opts;
  }

  public get awsTableName(): string {
    return this.tableName;
  }

  public get awsPartitionKey(): string {
    return "Id";
  }

  public async get(key: string): Promise<string> {
    const result = await this.docClient.send(
      new GetCommand({
        TableName: this.tableName,
        Key: {
          [this.awsPartitionKey]: key,
        },
      })
    );
    if (result.Item == undefined) {
      throw new Error(`There is no target key-value pair, Key: ${key}.`);
    }
    return result.Item["Value"];
  }

  public async set(key: string, val: string): Promise<void> {
    const command = new PutCommand({
      TableName: this.tableName,
      Item: {
        Id: key,
        Value: val,
      },
    });
    await this.docClient.send(command);
  }
}

File: packages/pluto-py/pluto_client/clients/aws/kvstore_dynamodb.py
from typing import Optional
import boto3
from pluto_base import utils
from ...kvstore import IKVStoreClient, KVStore, KVStoreOptions
from .utils import gen_aws_resource_name


class DynamoKVStore(IKVStoreClient):
    def __init__(self, name: str, opts: Optional[KVStoreOptions] = None):
        self.__id = utils.gen_resource_id(KVStore.fqn, name)
        self.__table_name = gen_aws_resource_name(self.__id)
        self.__client = boto3.resource("dynamodb").Table(self.__table_name)

    @property
    def aws_table_name(self) -> str:
        return self.__table_name

    @property
    def aws_partition_key(self) -> str:
        return "Id"

    def get(self, key: str) -> str:
        response = self.__client.get_item(Key={self.aws_partition_key: key})
        if "Item" not in response:
            raise ValueError(f"There is no target key-value pair, Key: {key}.")
        return response["Item"]["Value"]

    def set(self, key: str, val: str):
        self.__client.put_item(Item={"Id": key, "Value": val})
